noir-projects/aztec-nr/aztec/src/capsules/mod.nr

```diff
@@ -181,4 +217,78 @@ mod test {
 
         assert_eq(array.len(), 0);
     }
+
+    #[test]
+    unconstrained fn for_each_called_with_all_elements() {
+        let contract_address = setup();
+        let array = CapsuleArray::at(contract_address, SLOT);
+
+        array.push(4);
+        array.push(5);
+        array.push(6);
+
+        // We store all values that we were called with and check that all (value, index) tuples are present. Note that
+        // we do not care about the order in which each tuple was passed to the closure.
+        let called_with = &mut BoundedVec::<(u32, Field), 3>::new();
+        array.for_each(|index, value| { called_with.push((index, value)); });
+
+        assert_eq(called_with.len(), 3);
+        assert(called_with.any(|(index, value)| (index == 0) & (value == 4)));
+        assert(called_with.any(|(index, value)| (index == 1) & (value == 5)));
+        assert(called_with.any(|(index, value)| (index == 2) & (value == 6)));
+    }
+
+    #[test]
+    unconstrained fn for_each_remove_some() {
+        let contract_address = setup();
+        let array = CapsuleArray::at(contract_address, SLOT);
+
+        array.push(4);
+        array.push(5);
+        array.push(6);
+
+        array.for_each(|index, _| {
+            if index == 1 {
+                array.remove(index);
+            }
+        });
+
+        assert_eq(array.len(), 2);
+        assert_eq(array.get(0), 4);
+        assert_eq(array.get(1), 6);
+    }
+
+    #[test]
+    unconstrained fn for_each_remove_all() {
+        let contract_address = setup();
+        let array = CapsuleArray::at(contract_address, SLOT);
+
+        array.push(4);
+        array.push(5);
+        array.push(6);
+
+        array.for_each(|index, _| { array.remove(index); });
+
+        assert_eq(array.len(), 0);
+    }
+
+    // TODO: uncomment this test once OracleMock::count is implemented in the stdlib.
+    // #[test]
+    // unconstrained fn for_each_remove_all_no_copy() {
+    //     let contract_address = setup();
+    //     let array = CapsuleArray::at(contract_address, SLOT);
+
+    //     array.push(4);
+    //     array.push(5);
+    //     array.push(6);
+
+    //     // We test that the copyCapsule was never called, which is the expensive operation we want to avoid.
+    //     let mock = OracleMock::mock("copyCapsule");
+
+    //     array.for_each(|index, _| {
+    //         array.remove(index);
+    //     });
+
+    //     assert_eq(mock.count(), 0);
+    // }
```

Comment:
We'll be able uncomment this once https://github.com/noir-lang/noir/pull/7996 goes in and we get a sync PR.